### PR TITLE
Candace E

### PR DESCRIPTION
**Pull request by candace-eckels**
Completed User Story 1:
I'm pretty proud of how I handled my test anxiety this time around. I felt confident while coding and using the api calls along with the localhost to make sure I a returning what I am supposed to on the screen. 

I think I really need to focus on using the vcr to test the web page. I am really uncomfortable getting that setup on a time crunch. It really hinders me in edge case testing to not feel comfortable doing it under pressure.

I think if I had a little more time I would have tested my poros.

### CODE REVIEW

File: spec/features/search/index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
+require 'rails_helper'
+
+RSpec.describe 'Search Controller' do
```
Comment (dcoleman21) [inline spec/features/search/index_spec.rb:3]: This looks good. I would keep building on these tests and see about any sad paths. Like if you search something and it returns no results. You can try with your name.

File: app/views/search/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  <i>"<%= quote.content %>" </i> <br>
+  - <b> <%= quote.author %>, </b> 
+  tags: <% quote.tags.each do |tag| %>
+          <%= tag.titleize  %>
+        <% end %> <br>
+         <br>
+<% end %>
+<hr>
+
+Total Results: <%= number_with_delimiter(@result_count) %><br>
```
Comment (dcoleman21) [inline app/views/search/index.html.erb:18]: what is `number_with_delimiter` doing? When I spin up localhost, your total results are: 2,042. That seems like a lot. When I use Postman and search with the word `love` I am only getting a `:totalCount` of 122.

File: app/views/search/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+<br>
+<br>
+Searched for '<%= @query.titleize %>' Quotes
```
Comment (dcoleman21) [inline app/views/search/index.html.erb:3]: This all looks good.

File: app/services/quotable_service.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  def self.parse(response)
+    JSON.parse(response, symbolize_names: true)
+  end
+
+  def self.conn
+    Faraday.new(url: 'https://api.quotable.io')
+  end
+
+  def self.top_results(query)
+    response = conn.get("/quotes?query=#{query}&limit=10")
```
Comment (dcoleman21) [inline app/services/quotable_service.rb:15]: Nice job limiting here at the call to get the 10. 
I would look over the docs again. I think the endpoint that more closely gets what we are wanting is `/search/quotes`. Check this out https://github.com/lukePeavey/quotable#search-quotes-beta:~:text=einstein%7Cabraham%2Dlincoln-,Search%20Quotes%20(beta),-GET%20/search/quotes

File: app/services/quotable_service.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
+require 'faraday'
+require 'json'
+
+class QuotableService
```
Comment (dcoleman21) [inline app/services/quotable_service.rb:4]: Nice job making helpers for `parse and conn`

File: app/poros/quote.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+class Quote
+  attr_reader :author, :content, :tags
```
Comment (dcoleman21) [inline app/poros/quote.rb:2]: Looks good.

File: app/views/search/index.html.erb (hunk 2 above)
Comment (candace-eckels) [reply app/views/search/index.html.erb:18]: It adds a comma if it is a 4 digit number

File: app/poros/author.rb [hunk 7]
```diff
@@ -0,0 +1,3 @@
+class Author
```
Comment (dcoleman21) [inline app/poros/author.rb:1]: what were your thoughts for this file?

File: app/services/quotable_service.rb (hunk 4 above)
Comment (candace-eckels) [reply app/services/quotable_service.rb:15]: Ahhh you are right I do remeber reading that

File: app/poros/author.rb (hunk 7 above)
Comment (candace-eckels) [reply app/poros/author.rb:1]: If I made it to the extension I was going to create an author poro for author objects since I believe you can also search this api by authors as well It would be cool to be able to go to an author show page to view all of that persons quotes if they had multiple

File: app/facades/quotable_facade.rb [hunk 8]
```diff
@@ -0,0 +1,14 @@
+class QuotableFacade
+  def self.top_results(query)
```
Comment (dcoleman21) [inline app/facades/quotable_facade.rb:2]: Nice job taking into consideration that we want both the `:totalCount` and `quote attributes`. Here is a way that we can do that is a single method with a single call:
```
def self.quotes(query)
    json = QuotableService.top_results(query)
    quotes = json[:results].map do |quote_data|
      Quote.new(quote_data)
    end
    [json[:totalCount], quotes]
end
```
This will give us access to both things in a single call. And we can pass both of those to our controller as well.

File: app/controllers/search_controller.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+class SearchController < ApplicationController
+  def index
```
Comment (dcoleman21) [inline app/controllers/search_controller.rb:2]: See my comments in your Facade. If you were to implement that suggestion, your controller could look like:
```
def index
  @query = params[:query]
  @result_count, @top_results = QuotableFacade.top_results(@query)
end
```
Now your view has access to the word, the total count, and quote results for your view.

File: app/facades/quotable_facade.rb (hunk 8 above)
Comment (candace-eckels) [reply app/facades/quotable_facade.rb:2]: Ahhh! I was wondering if it made sense to do that in the same method or not! I should have went with my gut, because Why make two calls when I can just make one!

File: spec/facades/quotable_facade_spec.rb [hunk 10]
```diff
@@ -0,0 +1,25 @@
+require 'rails_helper'
+
```
Comment (dcoleman21) [inline spec/facades/quotable_facade_spec.rb:2]: Nice job testing here

File: spec/services/quotable_service_spec.rb [hunk 11]
```diff
@@ -0,0 +1,32 @@
+require 'rails_helper'
+
```
Comment (dcoleman21) [inline spec/services/quotable_service_spec.rb:2]: Nice job

File: app/controllers/search_controller.rb (hunk 9 above)
Comment (candace-eckels) [reply app/controllers/search_controller.rb:2]: Ohhh, How exactly does that `@result_count, @top_results = QuotableFacade.top_results(@query)` line work?

**Review by dcoleman21 — COMMENTED** [review]
Overall everything looks good. I would ask you to spin up your server and see if when you type in different keywords you get different results. When I had localhost running no matter what keyword I typed in, it returned the same results. I left you comments throughout to review. Let us know if you have any questions. Also let me know if you do indeed see different results with different keywords. I would also keep practicing sad path testing as well.
